### PR TITLE
Color picker/ Settings / Carousel View

### PR DESCRIPTION
**Pull request by s-fukuzawaa**
### Color Picker

- Installed new pod library for color picker
- Implemented markers to have personalized color
- Updated database to store user color
- Created color converter helper class to convert from hexString and colors (for the sake of saving to database)

### Settings
- Implemented update settings page
- Any updated fields will be saved to backend

### Carousel View
- Implemented carousel view for compose view
- Implemented carousel view for details view
- Implemented backend to save the images into Parse
- Started using PHPicker framework

- Demo
7F30C0B6-AC3F-4917-895F-CEC66BC996D2_2_0_a7F30C0B6-AC3F-4917-895F-CEC66BC996D2_2_0_a

### CODE REVIEW

File: Album/ViewControllers/ComposeViewController.m [hunk 1]
```diff
@@ -6,145 +6,215 @@
...
-	[self.pinImageView setImage:originalImage];
-	// TODO: Get date and set it to date picker default
-	NSURL *mediaUrl = info[UIImagePickerControllerMediaURL];
-	// Dismiss UIImagePickerController to go back to your original view controller
-	[self dismissViewControllerAnimated:YES completion:nil];
+- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *, id> *)info {
+    // Get the image captured by the UIImagePickerController
+    UIImage *originalImage = info[UIImagePickerControllerOriginalImage];
+    originalImage = [self resizeImage:originalImage withSize:self.pinImageView.image.size];
+    [self.pinImageView setImage:originalImage];
```
Comment (albertomota-codes) [inline Album/ViewControllers/ComposeViewController.m:136]:
```
[self.pinImageView setImage: [self resizeImage: info[UIImagePickerControllerOriginalImage]
                                                                 withSize:self.pinImageView.image.size]];
```

File: Album/ViewControllers/ComposeViewController.m [hunk 2]
```diff
@@ -6,145 +6,215 @@
...
+- (__kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
+    PhotoCollectionCell *photoCell = [collectionView dequeueReusableCellWithReuseIdentifier:@"photoCell" forIndexPath:indexPath];
+    // Add placeholder image cell when there are no images
+    if (self.photos.count == 0) {
+        photoCell.photoImageView.contentMode = UIViewContentModeScaleAspectFit;
+        photoCell.photoImageView.image = [UIImage imageNamed:@"image_placeholder"];
+    } else {
+        photoCell.photoImageView.contentMode = UIViewContentModeScaleAspectFit;
+        photoCell.photoImageView.image = self.photos[indexPath.row];
+    }
```
Comment (albertomota-codes) [inline Album/ViewControllers/ComposeViewController.m:208]:
```
photoCell.photoImageView.contentMode = UIViewContentModeScaleAspectFit;
photoCell.photoImageView.image = self.photos.count == 0 ? [UIImage imageNamed:@"image_placeholder"] : self.photos[indexPath.row];

File: Album/ViewControllers/MapControllers/FriendMapViewController.m [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+@property (nonatomic, strong) NSMutableArray *markerArr;
+@property (nonatomic, strong) NSMutableDictionary *placeToPins;
+@property (nonatomic, strong) NSMutableDictionary *pinImages;
+@property (nonatomic, strong) NSDateFormatter *formatter;
+@property (nonatomic, strong) ColorConvertHelper *colorHelper;
+@end
+
+@implementation FriendMapViewController
+
+- (void)loadView {
```
Comment (albertomota-codes) [inline Album/ViewControllers/MapControllers/FriendMapViewController.m:29]: Lets try to use `viewDidLoad`

Screen Shot 2022-07-22 at 10 19 11 AM

https://medium.com/good-morning-swift/ios-view-controller-life-cycle-2a0f02e74ff5

File: Album/ViewControllers/MapControllers/FriendMapViewController.m [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    self.locationManager.delegate = self;
+    // Ask for authentication
+    if ([self.locationManager respondsToSelector:@selector(requestWhenInUseAuthorization)]) {
+        [self.locationManager requestWhenInUseAuthorization];
+        [self.locationManager startUpdatingLocation];
+    }
+    // Set the intiial map view position
+    CLLocation *curPos = self.locationManager.location;
+    GMSCameraPosition *camera =
+    [GMSCameraPosition cameraWithLatitude:curPos.coordinate.latitude longitude:curPos.coordinate.longitude zoom:12];
```
Comment (albertomota-codes) [inline Album/ViewControllers/MapControllers/FriendMapViewController.m:45]: single line?

File: Album/ViewControllers/MapControllers/FriendMapViewController.m [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+    self.mapView.delegate = self;
+    // Set the date formatter
+    self.formatter = [[NSDateFormatter alloc] init];
+    [self.formatter setDateFormat:@"MMM dd, YYYY"];
+    [self.formatter setDateStyle:NSDateFormatterMediumStyle];
+    // Initialize data structures to cache retrieved data
+    self.placeToPins = [[NSMutableDictionary alloc] init];
+    self.pinImages = [[NSMutableDictionary alloc] init];
+} /* loadView */
+- (void)loadMarkers {
```
Comment (albertomota-codes) [inline Album/ViewControllers/MapControllers/FriendMapViewController.m:59]: space between 58 and 59

File: Album/ViewControllers/MapControllers/FriendMapViewController.m [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    PFQuery *query = [PFQuery queryWithClassName:classNameImage];
+    [query whereKey:@"pinId" equalTo:pinId];
+    return [query findObjects];
+}
+
+- (void)mapView:(GMSMapView *)mapView didTapInfoWindowOfMarker:(GMSMarker *)marker
+{
+    // If there are pins exist at this coordinate, lead to details otherwise compose view
+    if (self.placeToPins[marker.title]) {
+        [self performSegueWithIdentifier:@"detailsSegue" sender:marker];
```
Comment (albertomota-codes) [inline Album/ViewControllers/MapControllers/FriendMapViewController.m:214]: should the segue be in the Const file?

File: Album/ViewControllers/ProfileControllers/ActivitiesViewController.m [hunk 7]
```diff
@@ -17,66 +17,66 @@ @interface ActivitiesViewController ()<UITableViewDataSource, UITableViewDelegat
...
 }
 - (IBAction)backButton:(id)sender {
-	[self dismissViewControllerAnimated:YES completion:nil];
+    [self dismissViewControllerAnimated:YES completion:nil];
 }
 
-- (void) fetchActivities {
-	[self fetchRequests];
+- (void)fetchActivities {
+    [self fetchRequests];
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/ActivitiesViewController.m:38]: I would call it `fetchFriendRequest`

File: Album/ViewControllers/ProfileControllers/ActivitiesViewController.m [hunk 8]
```diff
@@ -17,66 +17,66 @@ @interface ActivitiesViewController ()<UITableViewDataSource, UITableViewDelegat
...
-	// Assign tableview data source and delegate
-	self.tableView.delegate = self;
-	self.tableView.dataSource = self;
-	self.requests = [[NSMutableArray alloc] init];
-	self.tableView.rowHeight = UITableViewAutomaticDimension;
+    [super viewDidLoad];
+    // Assign tableview data source and delegate
+    self.tableView.delegate = self;
+    self.tableView.dataSource = self;
+    self.requests = [[NSMutableArray alloc] init];
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/ActivitiesViewController.m:30]: I would call this `friendRequests`

File: Album/ViewControllers/ProfileControllers/ProfileViewController.m [hunk 9]
```diff
@@ -22,71 +22,67 @@ @interface ProfileViewController () <UICollectionViewDataSource, UICollectionVie
...
+    // Get the new view controller using [segue destinationViewController].
+    // Pass the selected object to the new view controller.
+   if ([segue.identifier isEqualToString:@"addFriendSegue"]) {
+        UINavigationController *navigationController = [segue destinationViewController];
+        AddFriendViewController *addFriendController = (AddFriendViewController *)navigationController.topViewController;
+        //        addFriendController.delegate = self; TODO: Add delegate later
+    } else if ([segue.identifier isEqualToString:@"settingsSegue"]) {
+        UINavigationController *navigationController = [segue destinationViewController];
+        SettingsViewController *settingsController = (SettingsViewController *)navigationController.topViewController;
+        //        settingsController.delegate = self; TODO: Add delegate later
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/ProfileViewController.m:84]: lets remove this?

File: Album/ViewControllers/ProfileControllers/SettingsViewController.m [hunk 10]
```diff
@@ -5,38 +5,205 @@
...
+        }];
+        // Add the upload from library action to the alert controller
+        [alert addAction:uploadAction];
+        //Cancel
+        UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"Cancel"
+                                                         style:UIAlertActionStyleDefault
+                                                       handler:nil];
+        // Add the cancel action to the alert controller
+        [alert addAction:cancel];
+        [self presentViewController:alert animated:YES completion:nil];
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/SettingsViewController.m:101]: LEts move the alerts to a helper function

**Review by albertomota-codes — APPROVED** [review]

File: Album/Helper/ColorConvertHelper.m [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+#import "ColorConvertHelper.h"
+
+@interface ColorConvertHelper ()
+
+@end
+
+@implementation ColorConvertHelper
+
+- (UIImage *)createImageWithColor:(UIColor *)color {
+    CGRect rect = CGRectMake(0.0f, 0.0f, 57, 57);
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/Helper/ColorConvertHelper.m:18]: Didn't know how to do this, will ask Monday

File: Album/Helper/ColorConvertHelper.m [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    UIGraphicsBeginImageContext(rect.size);
+    CGContextRef context = UIGraphicsGetCurrentContext();
+    CGContextSetFillColorWithColor(context, [color CGColor]);
+    CGContextFillRect(context, rect);
+
+    UIImage *theImage = UIGraphicsGetImageFromCurrentImageContext();
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/Helper/ColorConvertHelper.m:24]: Same as above

File: Album/Helper/ColorConvertHelper.m [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    [scanner scanHexInt:&rgbValue];
+    return [UIColor colorWithRed:((rgbValue & 0xFF0000) >>
+                                  16) / 255.0 green:((rgbValue & 0xFF00) >> 8) / 255.0 blue:(rgbValue & 0xFF) / 255.0 alpha:1.0];
+}
+
+- (NSString *)hexStringForColor:(UIColor *)color {
+    const CGFloat *components = CGColorGetComponents(color.CGColor);
+    CGFloat r = components[0];
+    CGFloat g = components[1];
+    CGFloat b = components[2];
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/Helper/ColorConvertHelper.m:41]: Done!

File: Album/ViewControllers/AuthenticationControllers/LoginViewController.m [hunk 14]
```diff
@@ -19,64 +19,64 @@ @interface LoginViewController ()
...
-		                            preferredStyle:(UIAlertControllerStyleAlert)];
-		// Create a cancel action
-		UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel"
-		                               style:UIAlertActionStyleCancel
-		                               handler:nil];
-		[alert addAction:cancelAction];
+    NSString *username = self.usernameField.text;
+    NSString *password = self.pwField.text;
+    // Alert for empty fields
+    if ([self.usernameField.text isEqual:@""] || [self.pwField.text isEqual:@""]) {
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/ViewControllers/AuthenticationControllers/LoginViewController.m:29]: I think so!

File: Album/ViewControllers/MapControllers/FriendMapViewController.m (hunk 6 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/MapControllers/FriendMapViewController.m:214]: Need clarification. Ask Monday!

File: Album/ViewControllers/ProfileControllers/ActivitiesViewController.m (hunk 7 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ProfileControllers/ActivitiesViewController.m:38]: Done!

File: Album/ViewControllers/ProfileControllers/ActivitiesViewController.m (hunk 8 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ProfileControllers/ActivitiesViewController.m:30]: Done!

File: Album/ViewControllers/ProfileControllers/ProfileViewController.m (hunk 9 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ProfileControllers/ProfileViewController.m:84]: Done!

File: Album/ViewControllers/ProfileControllers/SettingsViewController.m (hunk 10 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ProfileControllers/SettingsViewController.m:101]: Done!

File: Album/ViewControllers/ColorPickViewController.m [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+@interface ColorPickViewController () <FCColorPickerViewControllerDelegate>
+@property (nonatomic, strong) UIColor *color;
+@end
+
+@implementation ColorPickViewController
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    // Do any additional setup after loading the view.
+}
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/ViewControllers/ColorPickViewController.m:20]: Done!

File: Album/ViewControllers/ComposeViewController.m [hunk 16]
```diff
@@ -6,145 +6,215 @@
...
+    self.imageCarouselView.delegate = self;
+    self.imageCarouselView.dataSource = self;
+    self.imageCarouselView.layer.borderWidth = 1.0f;
+    self.imageCarouselView.layer.borderColor = [[UIColor blackColor] CGColor];
+    // Set up photos array
+    self.photos = [[NSMutableArray alloc] init];
+    self.currentIndex = 0;
+    // Set up page control
+    self.pageControl.numberOfPages = self.photos.count;
+} /* viewDidLoad */
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/ViewControllers/ComposeViewController.m:61]: It was automated using XCFormat

File: Album/ViewControllers/ComposeViewController.m [hunk 17]
```diff
@@ -6,145 +6,215 @@
...
-	[self.pinImageView addGestureRecognizer:imageTapGestureRecognizer];
-	[self.pinImageView setUserInteractionEnabled:YES];
+    [super viewDidLoad];
+    // Set caption view properties
+    self.captionTextView.delegate = self;
+    self.captionTextView.layer.cornerRadius = 8;
+    self.captionTextView.layer.borderWidth = 1.0f;
+    self.captionTextView.placeholder = @"Add a caption...";
+    self.captionTextView.placeholderColor = [UIColor lightGrayColor];
+    self.captionTextView.layer.borderColor = [[UIColor systemBlueColor] CGColor];
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/ViewControllers/ComposeViewController.m:39]: Done!

File: Album/ViewControllers/ComposeViewController.m [hunk 18]
```diff
@@ -6,145 +6,215 @@
...
+    self.locationLabel.text = [self.locationLabel.text stringByAppendingString:self.placeName];
+    // Config PHPicker
+    self.config = [[PHPickerConfiguration alloc] init];
+    self.config.selectionLimit = 10;
+    self.config.filter = [PHPickerFilter imagesFilter];
+    // Photo carousel
+    self.imageCarouselView.delegate = self;
+    self.imageCarouselView.dataSource = self;
+    self.imageCarouselView.layer.borderWidth = 1.0f;
+    self.imageCarouselView.layer.borderColor = [[UIColor blackColor] CGColor];
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/ViewControllers/ComposeViewController.m:55]: Done!

File: Album/ViewControllers/ComposeViewController.m [hunk 19]
```diff
@@ -6,145 +6,215 @@
...
+        }
+    }];
+    
+    [self returnMap];
+} /* postButton */
+- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
+    if (self.photos.count == 0) {
+        return 1;
+    }
+    return self.photos.count;
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/ViewControllers/ComposeViewController.m:195]: Done!

File: Album/ViewControllers/ComposeViewController.m [hunk 20]
```diff
@@ -6,145 +6,215 @@
...
+            NSLog(@"Error posting: %@", error.localizedDescription);
+        } else {
+            NSLog(@"Pin saved successfully! Object Id:%@", newPin.objectId);
+            for (UIImage *image in self.photos) {
+                [Image postImage:image withPin:newPin.objectId withCompletion:^(BOOL succeeded, NSError *_Nullable error) {
+                    if (error) {
+                        NSLog(@"Error saving image: %@", error.localizedDescription);
+                    } else {
+                        NSLog(@"Successfully saved image");
+                    }
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/ViewControllers/ComposeViewController.m:182]: Done!

File: Album/ViewControllers/ComposeViewController.m (hunk 1 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ComposeViewController.m:136]: Done!

File: Album/ViewControllers/ComposeViewController.m (hunk 2 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ComposeViewController.m:208]: Done!

File: Album/ViewControllers/MapControllers/FriendMapViewController.m (hunk 4 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/MapControllers/FriendMapViewController.m:45]: Done!

File: Album/ViewControllers/MapControllers/FriendMapViewController.m (hunk 5 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/MapControllers/FriendMapViewController.m:59]: Done!

File: Album/ViewControllers/MapControllers/FriendMapViewController.m (hunk 6 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/MapControllers/FriendMapViewController.m:214]: Done!

File: Album/Helper/ColorConvertHelper.m (hunk 11 above)
[the remark this answers is not in the document]
Comment (albertomota-codes) [reply Album/Helper/ColorConvertHelper.m]: `const CGRect rect = CGRectMake(0.0f, 0.0f, 57, 57)`

File: Album/Helper/ColorConvertHelper.m (hunk 12 above)
[the remark this answers is not in the document]
Comment (albertomota-codes) [reply Album/Helper/ColorConvertHelper.m:24]: `UIImage *const rectangleFilledImage = UIGraphicsGetImageFromCurrentImageContext();`

**Pull request merged**